kernel/src/table_configuration.rs

```diff
@@ -176,29 +181,199 @@ impl TableConfiguration {
+                // Legacy reader: infer features from version
+                LEGACY_READER_FEATURES
+                    .iter()
+                    .filter(|f| {
+                        f.info()
+                            .map(|info| v >= info.min_reader_version)
+                            .unwrap_or(false)
+                    })
+                    .cloned()
+                    .collect()
+            }
+            _ => Vec::new(),
+        }
+    }
+
+    /// Returns all writer features enabled for this table based on protocol version.
+    /// For table features protocol (v7), returns the explicit writer_features list.
+    /// For legacy protocol (v1-6), infers features from the version number.
+    fn get_enabled_writer_features(&self) -> Vec<TableFeature> {
+        match self.protocol.min_writer_version() {
+            7 => {
+                // Table features writer: use explicit writer_features list
+                self.protocol
+                    .writer_features()
+                    .map(|f| f.to_vec())
+                    .unwrap_or_default()
+            }
+            v if (1..=6).contains(&v) => {
+                // Legacy writer: infer features from version
+                LEGACY_WRITER_FEATURES
+                    .iter()
+                    .filter(|f| {
+                        f.info()
+                            .map(|info| v >= info.min_writer_version)
+                            .unwrap_or(false)
+                    })
+                    .cloned()
+                    .collect()
+            }
+            _ => Vec::new(),
+        }
+    }
+
+    /// Returns `Ok` if the kernel supports reading from this table. This checks that the
+    /// protocol's reader features are all supported.
+    fn ensure_read_supported(&self) -> DeltaResult<()> {
+        // Version check
+        match self.protocol.min_reader_version() {
+            1..=3 => {}
+            _ => {
+                return Err(Error::unsupported(format!(
+                    "Unsupported minimum reader version {}",
+                    self.protocol.min_reader_version()
+                )))
+            }
+        }
+
+        // Check all enabled reader features have kernel read support
+        for feature in self.get_enabled_reader_features() {
+            let Some(info) = feature.info() else {
+                return Err(Error::unsupported(format!("Unknown feature '{}'", feature)));
+            };
+
+            // Check read support
+            match &info.read_support {
+                KernelSupport::Supported => Ok(()),
+                KernelSupport::NotSupported => Err(Error::unsupported(format!(
+                    "Feature '{}' not supported for reads",
+                    info.name
+                ))),
+                KernelSupport::Custom(check) => {
+                    check(&self.protocol, &self.table_properties, Operation::Scan)
+                }
+            }?;
```

Comment:
That `?` operator is sneaky. Let's just inline it
```suggestion
            match &info.read_support {
                KernelSupport::Supported => {},
                KernelSupport::NotSupported => return Err(Error::unsupported(format!(
                    "Feature '{}' not supported for reads",
                    info.name
                ))),
                KernelSupport::Custom(check) => {
                    check(&self.protocol, &self.table_properties, Operation::Scan)?;
                }
            };
```